src/inatinqperf/utils/profiler.py

```diff
@@ -5,12 +5,90 @@
 import os
 import time
 import tracemalloc
+from collections.abc import Sequence
+from dataclasses import dataclass, field
 from pathlib import Path
+from statistics import mean
 from types import TracebackType
+from typing import Any, Self
 
 import psutil
 from loguru import logger
 
+try:  # Optional dependency used only when container tracking is enabled.
+    import docker
+    from docker.models.containers import Container
+except ImportError:  # pragma: no cover - exercised when docker is unavailable
+    docker = None
+    Container = Any  # type: ignore[misc]
+
+
+@dataclass(slots=True)
+class _ContainerTracker:
+    """Collect CPU and memory metrics for a Docker container."""
+
+    container: Container
+    cpu_samples: list[float] = field(default_factory=list)
+    memory_samples: list[float] = field(default_factory=list)
+
+    def sample(self) -> None:
+        """Capture a stats snapshot, ignoring transient failures."""
+        try:
+            stats = self.container.stats(stream=False)
+        except Exception:  # pragma: no cover - relies on docker API behaviour
+            logger.opt(exception=True).debug("[PROFILE] Failed to gather container stats")
+            return
+
+        try:
+            memory_usage = stats.get("memory_stats", {}).get("usage")
+            if memory_usage is not None:
+                self.memory_samples.append(memory_usage / (1024 * 1024))
+
+            cpu_percent = self._cpu_percent(stats)
+            if cpu_percent is not None:
+                self.cpu_samples.append(cpu_percent)
+        except Exception:  # pragma: no cover - defensive against schema drift
+            logger.opt(exception=True).debug("[PROFILE] Unexpected container stats payload")
+
+    def summary(self) -> dict[str, Any]:
```

Comment:
Prefer `object` type hint over `Any`, as `Any` tells the type checker to not look at the values at all, forever.